db/migrate/20190102161442_add_taggings_for_all_taggable_tables.rb

```diff
@@ -0,0 +1,56 @@
+class AddTaggingsForAllTaggableTables < ActiveRecord::Migration[5.1]
+  def change
+    create_table "tags", id: :serial, force: :cascade do |t|
+      t.references :tenant, :type => :bigint, :index => false, :null => false, :foreign_key => {:on_delete => :cascade}
+
+      t.string "name", :null => false
+      t.text "description"
+      t.boolean "active", :default => true, :null => false
+
+      # TODO what is human readable name? Will we use it for searching? How will we search with translated field?
+      t.string "display_name"
+
+      # TODO the queries like "starts_with" won't be consistent across platform, if we'll model separately vs. encoded
+      # as namespace/name. Or we'll have to do query transformations, will we do that?
+      t.string "namespace"
```

Comment:
@agrare @Fryguy please advise, if we'll model it separately while IPP says it should be prefix, we'l need to transform a lot of queries, e.g. for ```aws/env```:

```starts_with("aws/e") => name.starts_with("aws") && namespace.starts_with("e")```
```contains("/env") => namespace IS NOT NULL && name.contains("env")```

...
so should we rather model it as prefix of the name?

Reply:
Good point, either way core should handle setting the namespace (either in the tag name or separate) rather than the provider author.

We'll need to split or join the namespace at some point.  If we store it separately we'll need to split what is passed in when doing a query, and if we store it in the tag name we'll need to split it to do any native operations like tagging a vm.

> The key syntax will be a UTF-8 encoded string containing only letters, numbers, hyphens, and underscores. Optionally, a namespace prefix may be attached to the key name and delimited with a slash. 

This sounds like we'd be able to assume that any slash delimits the namespace and the rest of the tag but I'm not sure we can assume that native source tags won't have `'/'` in them.

Reply:
Hm, I've leaned to have it encoded to the name, but your right, we can't ensure provider won't have '/' in the tag name. So it looks like it should be in a separate column after all.

Reply:
Also it looks like the namespace should really be associated with source_type. So it gets assigned automatically. Should we rather associate source or source_type here, instead of namespace?

Reply:
Actually we'd probably have an easier time stripping it off the name in the operations worker because we already know the source_type so we could strip exactly e.g. `"aws/"` instead of just trying to match on the first `"/"` in the query

Reply:
Maybe. If you'd call you provider tag as ```aws/env```, we'd store it as ```aws/aws/env```. So it would still work I guess :-)